hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsCompressor.java

```diff
@@ -257,29 +259,26 @@ private void writeDictionary() throws IOException {
         dictionaryDos.writeInt(words.size());
         String lastWord = "";
         for (MetricsDictionary.Word word : words) {
-
             String wordText = word.word();
-            int maxCommonLen = Math.min(lastWord.length(), wordText.length());
+            if (wordText.length() > UNSIGNED_BYTE_MAX_VALUE) {
+                // this should have been checked earlier, this is a safety check
+                throw new RuntimeException("Dictionary element too long: " + wordText);
+            }
+            int maxCommonLen = Math.min(lastWord.length(), wordText.length()) & UNSIGNED_BYTE_MAX_VALUE;
             int commonLen = 0;
-            boolean common = true;
-            for (int i = 0; i < maxCommonLen && common; i++) {
-                if (wordText.charAt(i) == lastWord.charAt(i)) {
-                    commonLen++;
-                } else {
-                    common = false;
-                }
+            while (commonLen < maxCommonLen
+                    && wordText.charAt(commonLen) == lastWord.charAt(commonLen)) {
+                commonLen++;
             }
 
             int diffLen = wordText.length() - commonLen;
-            tmpDos.writeInt(word.dictionaryId());
-            tmpDos.writeByte(commonLen);
-            tmpDos.writeByte(diffLen);
+            dictionaryDos.writeInt(word.dictionaryId());
```

Comment:
I believe we should keep writing to `tmpDos` and copy its backing buffer to `dictionaryDos` at the end. The reason for introducing `tmpDos` was to avoid allocating many single-byte `byte[]` instances that `DeflaterOutputStream` does for writing (most of the) primitives. See `DeflaterOutputStream#write(int)`.

Reply:
That's nasty.. I'm reverting then and adding a comment. A code that takes a more complicated path for simple things for some side-effect should have a comment IMO.

Reply:
> A code that takes a more complicated path for simple things for some side-effect should have a comment IMO.

The comment is already there:
https://github.com/hazelcast/hazelcast/blob/97ff57ba92dc1fbfa77720885e7db40757999080/hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsCompressor.java#L124

Reply:
Renaming the field to `allocFriendlyDos` or something meaningful would also make sense. That should at least give a hint to future us. `tmpDos` is indeed a meaningless name and makes us thinking using this field is not necessary. I had to navigate to the field declaration as well to recall the reason for using it.